File: src/dantofa/core/digitalocean/spaces.py
"""Application logic for DigitalOcean Spaces bucket management.

Framework-free. The Spaces (S3) surface is reached through a structurally-typed
:class:`SupportsSpacesApi` client; the concrete boto3 adapter lives in
``dantofa.clients.digitalocean.spaces``.
"""

from __future__ import annotations

from typing import Protocol


class SupportsSpacesApi(Protocol):
    """The Spaces-API surface this module depends on (see clients adapter)."""

    def list(self) -> list[dict[str, object]]: ...
    def create(self, name: str) -> None: ...
    def delete(self, name: str) -> None: ...


def list_buckets(client: SupportsSpacesApi) -> list[dict[str, object]]:
    return client.list()


def create_bucket(client: SupportsSpacesApi, name: str) -> None:
    """Create a bucket. Surfaces the raw S3 error if it already exists."""
    client.create(name)


def delete_bucket(client: SupportsSpacesApi, name: str) -> None:
    # SKY-D216 is a false positive here: delete() is a bucket-name Protocol call,
    # not a URL/HTTP sink — there is no SSRF surface in this module.
    client.delete(name)  # skylos: ignore[SKY-D216]

File: internal/core/digitalocean/spaces.go
// Package digitalocean holds the framework-free application logic for
// DigitalOcean resources (Spaces buckets and DOKS clusters). It imports neither
// a CLI framework nor the client SDKs: the provider surfaces are reached through
// the interfaces declared here, which the clients adapter satisfies.
package digitalocean

import (
	"context"
	"time"
)

// Bucket is a Spaces bucket as surfaced to the user.
type Bucket struct {
	Name      string    `json:"name"`
	CreatedAt time.Time `json:"created_at"`
}

// SpacesAPI is the Spaces (S3) surface this package depends on.
type SpacesAPI interface {
	ListBuckets(ctx context.Context) ([]Bucket, error)
	CreateBucket(ctx context.Context, name string) error
	DeleteBucket(ctx context.Context, name string) error
}

// ListBuckets returns all Spaces buckets.
func ListBuckets(ctx context.Context, client SpacesAPI) ([]Bucket, error) {
	return client.ListBuckets(ctx)
}

// CreateBucket creates a bucket, surfacing the raw provider error if it already
// exists.
func CreateBucket(ctx context.Context, client SpacesAPI, name string) error {
	return client.CreateBucket(ctx, name)
}

// DeleteBucket deletes a bucket.
func DeleteBucket(ctx context.Context, client SpacesAPI, name string) error {
	return client.DeleteBucket(ctx, name)
}
